internal/app/server/impl.go

```diff
@@ -0,0 +1,374 @@
+// Copyright 2023 The Connect Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"time"
+
+	"connectrpc.com/conformance/internal/gen/proto/connect/connectrpc/conformance/v1alpha1/conformancev1alpha1connect"
+	v1alpha1 "connectrpc.com/conformance/internal/gen/proto/go/connectrpc/conformance/v1alpha1"
+	connect "connectrpc.com/connect"
+	proto "google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/anypb"
+)
+
+// ConformanceRequest is a general interface for all conformance requests (UnaryRequest, ServerStreamRequest, etc.)
+type ConformanceRequest interface {
+	GetResponseHeaders() []*v1alpha1.Header
+	GetResponseTrailers() []*v1alpha1.Header
+}
+
+// Stream represents a stream with the ability to set headers and trailers and send a message T
+type Stream[T any] interface {
+	ResponseHeader() http.Header
+	ResponseTrailer() http.Header
+	Send(*T) error
+}
+
+type conformanceServer struct{}
+
+func (s *conformanceServer) Unary(
+	ctx context.Context,
+	req *connect.Request[v1alpha1.UnaryRequest],
+) (*connect.Response[v1alpha1.UnaryResponse], error) {
+	msgAsAny, err := asAny(req.Msg)
+	if err != nil {
+		return nil, err
+	}
+	payload, err := parseUnaryResponseDefinition(req.Msg.ResponseDefinition, req.Header(), []*anypb.Any{msgAsAny})
+	if err != nil {
+		return nil, err
```

Comment:
You still need to populate response metadata. For unary RPCs, this is done by adding header metadata to a `*connect.Error`. It has a `Meta()` method that returns a mutable `http.Header`.

Sadly, the API provides no way to distinguish between headers and trailers in the error metadata. I think this is because, in practice, errors from unary RPCs often result in "trailers-only responses" in gRPC, which also cannot distinguish.

So I would do something like this here:
```suggestion
		addHeaders(req.Msg.ResponseDefinition.ResponseHeaders, err.Meta())
		addHeaders(req.Msg.ResponseDefinition.ResponseTrailers, err.Meta())
		return nil, err
```

Reply:
I added this to unary and client stream requests. Should this also be included on the server and bidi requests in the event we return with an error?

Reply:
Nah. For those, the `stream` type the handler gets has ways to set headers and trailers separately. If you _do_ add metadata to the error, it will get merged into the trailers. But I don't think we want or need to do that here for those endpoints.